### ISSUE TITLE
Unexpected return type for `all()`

### ISSUE THREAD

**Issue by bdougherty** [thefrontside/effection#870]
Say I have this code:

```ts
import { sleep, main, all, type Operation } from 'effection';

function* slowOperation<T>(value: T): Operation<T> {
    yield* sleep(1000);
    return value;
}

await main(function* () {
    const [result1, result2, result3] = yield* all([slowOperation('string'), slowOperation(5)]);
    console.log(result1.split('r'));
    console.log(result2.toFixed(3));
});
```

TypeScript will give an error on both of those console logs because both `result1` and `result2` have the type of `string | number`. It will also miss the error on `result3`, which will never exist here.

The return type of `yield* all()` in this example should be `[string, number]`, not `(string | number)[]`, unless I am missing something or doing something wrong. The `Promise.all()` return types are as I would expect.

**Comment by cowboyd** [comment]
@bdougherty Thank you for reporting this!. We should definitely be emulating the typing in Promise.all

**Comment by cowboyd** [comment]
It looks like there are overrides for up to 10 promises to support this https://github.com/microsoft/TypeScript/blob/main/src/lib/es2015.promise.d.ts

**Comment by cowboyd** [comment]
@bdougherty Are there any other test cases that you think are missing?

https://github.com/thefrontside/effection/blob/31e55bf95f14f35d548e77b18a099a33a5f510cf/test/all.test.ts#L97-L106)

**Comment by bdougherty** [comment]
Only thing that comes to mind is arbitrary array lengths, but it would surprise me if that wasn't covered by those tests.

**Comment by taras** [comment]
@bdougherty, the fix for these types was released today in 3.0.1. Give it a shot.
